typescript/infra/src/tx/transaction-reader.ts

```diff
@@ -0,0 +1,489 @@
+        info: 'Incorrect destination ICA in ICA call',
+      });
+      remoteIcaInsight = `❌ fatal mismatch, expected ${remoteIcaAddress}`;
+    }
+
+    const decodedCalls = await Promise.all(
+      calls.map((call: any) => {
+        const icaCallAsTx = {
+          to: bytes32ToAddress(call[0]),
+          value: BigNumber.from(call[1]),
+          data: call[2],
+        };
+        return this.read(remoteChainName, icaCallAsTx);
+      }),
+    );
+
+    return {
+      destination: {
+        domain: destination,
+        chainName: remoteChainName,
+      },
+      router: {
+        address: router,
+        insight: routerInsight,
+      },
+      ism: {
+        address: ism,
+        insight: ismInsight,
+      },
+      destinationIca: {
+        address: remoteIcaAddress,
+        insight: remoteIcaInsight,
+      },
+      calls: decodedCalls,
+    };
+  }
+
+  private async readMultisendTransaction(
+    chain: ChainName,
+    tx: AnnotatedEV5Transaction,
+  ): Promise<any> {
+    if (!tx.data) {
+      console.log('No data in multisend transaction');
+      return undefined;
+    }
+    const multisends = decodeMultiSendData(tx.data);
+
+    const { symbol } = await this.multiProvider.getNativeToken(chain);
+
+    return Promise.all(
+      multisends.map(async (multisend, index) => {
+        const decoded = await this.read(
+          chain,
+          metaTransactionDataToEV5Transaction(multisend),
+        );
+        return {
+          index,
+          value: `${ethers.utils.formatEther(multisend.value)} ${symbol}`,
+          operation: formatOperationType(multisend.operation),
+          decoded,
+        };
+      }),
+    );
+  }
+
+  isIcaTransaction(chain: ChainName, tx: AnnotatedEV5Transaction): boolean {
+    return (
+      tx.to !== undefined &&
+      eqAddress(tx.to, this.chainAddresses[chain].interchainAccountRouter)
+    );
+  }
+
+  isMailboxTransaction(chain: ChainName, tx: AnnotatedEV5Transaction): boolean {
+    return (
+      tx.to !== undefined &&
+      eqAddress(tx.to, this.chainAddresses[chain].mailbox)
+    );
+  }
+
+  async isMultisendTransaction(
+    chain: ChainName,
+    tx: AnnotatedEV5Transaction,
+  ): Promise<boolean> {
+    if (tx.to === undefined) {
+      return false;
+    }
+    const multiSendCallOnlyAddress = await this.getMultiSendCallOnlyAddress(
+      chain,
+    );
+    if (!multiSendCallOnlyAddress) {
+      return false;
+    }
+
+    return eqAddress(multiSendCallOnlyAddress, tx.to);
+  }
+
+  multiSendCallOnlyAddressCache: ChainMap<string> = {};
+
+  async getMultiSendCallOnlyAddress(
+    chain: ChainName,
+  ): Promise<string | undefined> {
+    if (this.multiSendCallOnlyAddressCache[chain]) {
+      return this.multiSendCallOnlyAddressCache[chain];
+    }
+
+    const safe = safes[chain];
+    if (!safe) {
+      return undefined;
+    }
+
+    const { safeSdk } = await retryAsync(() =>
+      getSafeAndService(chain, this.multiProvider, safe),
```

Comment:
can move the retryAsync logic into `getSafeAndService` directly - think it's only required when doing the `getSafe` call?
```ts
export async function getSafeAndService(
  chain: ChainNameOrId,
  multiProvider: MultiProvider,
  safeAddress: Address,
) {
  const safeSdk: Safe.default = await retryAsync(
    () => getSafe(chain, multiProvider, safeAddress),
    5,
    1000,
  );
  const safeService: SafeApiKit.default = getSafeService(chain, multiProvider);
  return { safeSdk, safeService };
}
```